frontend/src/views/admin/matching/index.tsx

```diff
@@ -21,11 +83,218 @@ export function AdminMatchingView() {
+            ApplicantSummary[]
+        > = {};
+        for (const applicant of applicants) {
+            const newestApplication = GetNewestApplication(
+                applicant,
+                applications
+            );
+            if (!newestApplication) {
+                continue;
+            }
+
+            const applicantMatches =
+                matches.filter((match) => match.applicantId === applicant.id) ||
+                [];
+
+            const applicantGuarantee =
+                guarantees.find(
+                    (guarantee) => guarantee.utorid === applicant.utorid
+                ) || null;
+
+            const newApplicantSummary = {
+                applicant: applicant,
+                mostRecentApplication: newestApplication,
+                matches: applicantMatches,
+                guarantee: applicantGuarantee,
+            };
+
+            newApplicantSummary.mostRecentApplication.position_preferences.forEach(
+                (position) => {
+                    applicantSummariesByPositionId[position.position.id] =
+                        applicantSummariesByPositionId[position.position.id] ||
+                        [];
+                    applicantSummariesByPositionId[position.position.id].push(
+                        newApplicantSummary
+                    );
+                }
+            );
+
+            // Add summary to positions where the applicant has been assigned:
+            assignments
+                .filter(
+                    (assignment) => assignment.applicant.id === applicant.id
+                )
+                .forEach((assignment) => {
+                    applicantSummariesByPositionId[assignment.position.id] =
+                        applicantSummariesByPositionId[
+                            assignment.position.id
+                        ] || [];
+
+                    const existingSummary = applicantSummariesByPositionId[
+                        assignment.position.id
+                    ].find((summary) => summary.applicant.id === applicant.id);
+
+                    // Add the applicant summary to the position only if the summary does not already exist
+                    if (!existingSummary) {
+                        applicantSummariesByPositionId[
+                            assignment.position.id
+                        ].push(newApplicantSummary);
+                    }
+                });
+        }
+
+        const ret: Record<number, PositionSummary> = {};
+        for (const position of positions) {
+            const targetHours = round(
+                position.hours_per_assignment *
+                    (position.desired_num_assignments || 0),
+                2
+            );
+
+            let hoursAssigned = 0;
+            for (const match of matches.filter(
+                (match) =>
+                    ["assigned", "staged-assigned"].includes(match.status) &&
+                    match.positionId === position.id
```

Comment:
Quadratic behavior again? I feel like you might be able to assemble the data you want more cleanly if you stepped back and thought about it again.